Lib/random.py

```diff
@@ -323,44 +323,56 @@ def sample(self, population, k):
         large population:   sample(range(10000000), 60)
         """
 
-        # Sampling without replacement entails tracking either potential
-        # selections (the pool) in a list or previous selections in a set.
-
-        # When the number of selections is small compared to the
-        # population, then tracking selections is efficient, requiring
-        # only a small set and an occasional reselection.  For
-        # a larger number of selections, the pool tracking method is
-        # preferred since the list takes less space than the
-        # set and it doesn't suffer from frequent reselections.
-
-        if isinstance(population, _Set):
-            population = tuple(population)
-        if not isinstance(population, _Sequence):
+        # Where k is n/2 or more, or where the population is a set and so we
+        # have to iterate through every item anyway, use reservoir sampling
+        # (Knuth's "Algorithm R").
+
+        # Otherwise use the "cardchoose" algorithm: select fairly from
+        # all the possible sorted lists of integers 0 <= item < n - k + 1
+        # (for this we have to introduce a bias in favour of duplicates,
+        # which the sort fixes) then turn this into a random choice from
+        # the population. Consider eg random.randchoose(range(3), 2):
+        # then n - k + 1 = 2 and the possible sorted lists are
+        # [0, 0], [0, 1], [1, 1], each of which can arise in two different ways.
+        # After `result[i] += i` this becomes [0, 1], [0, 2], or [1, 2];
```

Comment:
If you don't read the implementation `result[i]` makes no sense. There is no previous mention of what that collection is representing.